File: wire_encode.go
package ratify

// Wire encoders — the official way to turn protocol structures into their
// canonical wire JSON (SPEC §5.7, §5.8, §16.3): lex-sorted keys, byte
// fields as base64-standard strings, optional v1.1 fields omitted when
// empty. All three route through CanonicalJSON, so integer fields are
// validated against the safe-integer domain (SPEC §6.2) and the encoder
// returns an error rather than emitting a document strict wire decoders
// reject. Direct json.Marshal of the exported structs would emit fields in
// declaration order, not canonical order — always encode wire documents
// through these helpers.

// delegationCertWire mirrors DelegationCert with fields in alphabetical
// JSON-key order (Go marshals in declaration order) and includes the
// signature — unlike delegationSignable, which is the signed subset.
type delegationCertWire struct {
	CertID        string          `json:"cert_id"`
	Constraints   []Constraint    `json:"constraints"`
	ExpiresAt     int64           `json:"expires_at"`
	IssuedAt      int64           `json:"issued_at"`
	IssuerID      string          `json:"issuer_id"`
	IssuerPubKey  HybridPublicKey `json:"issuer_pub_key"`
	Scope         []string        `json:"scope"`
	Signature     HybridSignature `json:"signature"`
	SubjectID     string          `json:"subject_id"`
	SubjectPubKey HybridPublicKey `json:"subject_pub_key"`
	Version       int             `json:"version"`
}

type proofBundleWire struct {
	AgentID        string               `json:"agent_id"`
	AgentPubKey    HybridPublicKey      `json:"agent_pub_key"`
	Challenge      []byte               `json:"challenge"`
	ChallengeAt    int64                `json:"challenge_at"`
	ChallengeSig   HybridSignature      `json:"challenge_sig"`
	Delegations    []delegationCertWire `json:"delegations"`
	SessionContext []byte               `json:"session_context,omitempty"`
	StreamID       []byte               `json:"stream_id,omitempty"`
	StreamSeq      int64                `json:"stream_seq,omitempty"`
}

type sessionTokenWire struct {
	AgentID      string          `json:"agent_id"`
	AgentPubKey  HybridPublicKey `json:"agent_pub_key"`
	ChainHash    []byte          `json:"chain_hash"`
	GrantedScope []string        `json:"granted_scope"`
	HumanID      string          `json:"human_id"`
	IssuedAt     int64           `json:"issued_at"`
	MAC          []byte          `json:"mac"`
	SessionID    string          `json:"session_id"`
	ValidUntil   int64           `json:"valid_until"`
	Version      int             `json:"version"`
}

func delegationCertWireOf(cert *DelegationCert) delegationCertWire {
	constraints := cert.Constraints
	if constraints == nil {
		// Canonical form serializes constraints as [] when empty, never null.
		constraints = []Constraint{}
	}
	return delegationCertWire{
		CertID:        cert.CertID,
		Constraints:   constraints,
		ExpiresAt:     cert.ExpiresAt,
		IssuedAt:      cert.IssuedAt,
		IssuerID:      cert.IssuerID,
		IssuerPubKey:  cert.IssuerPubKey,
		Scope:         cert.Scope,
		Signature:     cert.Signature,
		SubjectID:     cert.SubjectID,
		SubjectPubKey: cert.SubjectPubKey,
		Version:       cert.Version,
	}
}

// EncodeDelegationCert marshals cert into its canonical wire JSON
// (SPEC §5.7). Integer fields outside the safe-integer domain are an
// error, never emitted.
func EncodeDelegationCert(cert *DelegationCert) ([]byte, error) {
	w := delegationCertWireOf(cert)
	return CanonicalJSON(w)
}

// EncodeProofBundle marshals bundle into its canonical wire JSON
// (SPEC §5.8). Optional v1.1 fields (session_context, stream_id,
// stream_seq) are omitted when empty, matching the reference wire form.
// Integer fields outside the safe-integer domain are an error, never
// emitted.
func EncodeProofBundle(bundle *ProofBundle) ([]byte, error) {
	delegations := make([]delegationCertWire, len(bundle.Delegations))
	for i := range bundle.Delegations {
		delegations[i] = delegationCertWireOf(&bundle.Delegations[i])
	}
	w := proofBundleWire{
		AgentID:        bundle.AgentID,
		AgentPubKey:    bundle.AgentPubKey,
		Challenge:      bundle.Challenge,
		ChallengeAt:    bundle.ChallengeAt,
		ChallengeSig:   bundle.ChallengeSig,
		Delegations:    delegations,
		SessionContext: bundle.SessionContext,
		StreamID:       bundle.StreamID,
		StreamSeq:      bundle.StreamSeq,
	}
	return CanonicalJSON(w)
}

// EncodeSessionToken marshals token into its canonical wire JSON
// (SPEC §16.3). Integer fields outside the safe-integer domain are an
// error, never emitted.
func EncodeSessionToken(token *SessionToken) ([]byte, error) {
	w := sessionTokenWire{
		AgentID:      token.AgentID,
		AgentPubKey:  token.AgentPubKey,
		ChainHash:    token.ChainHash,
		GrantedScope: token.GrantedScope,
		HumanID:      token.HumanID,
		IssuedAt:     token.IssuedAt,
		MAC:          token.MAC,
		SessionID:    token.SessionID,
		ValidUntil:   token.ValidUntil,
		Version:      token.Version,
	}
	return CanonicalJSON(w)
}

File: sdks/rust/tests/wire_encode.rs
// Encoder side of the wire integer domain (SPEC §6.2): serialization of
// wire structures must never emit an integer that strict wire decoders
// reject. Positive cases sit exactly on the safe-integer bounds.

use ratify_protocol::{
    Constraint, DelegationCert, HybridPublicKey, HybridSignature, ProofBundle, SessionToken,
};

const MAX_SAFE: i64 = (1 << 53) - 1;

fn pubkey() -> HybridPublicKey {
    HybridPublicKey {
        ed25519: vec![0; 32],
        ml_dsa_65: vec![0; 1952],
    }
}

fn sig() -> HybridSignature {
    HybridSignature {
        ed25519: vec![0; 64],
        ml_dsa_65: vec![0; 3309],
    }
}

fn token(issued_at: i64, valid_until: i64) -> SessionToken {
    SessionToken {
        version: 1,
        session_id: "s".into(),
        agent_id: "a".into(),
        agent_pub_key: pubkey(),
        human_id: "h".into(),
        granted_scope: vec![],
        issued_at,
        valid_until,
        chain_hash: vec![0; 32],
        mac: vec![0; 32],
    }
}

fn cert(issued_at: i64, expires_at: i64, constraints: Vec<Constraint>) -> DelegationCert {
    DelegationCert {
        cert_id: "c".into(),
        version: 1,
        issuer_id: "i".into(),
        issuer_pub_key: pubkey(),
        subject_id: "s".into(),
        subject_pub_key: pubkey(),
        scope: vec!["meeting:attend".into()],
        constraints,
        issued_at,
        expires_at,
        signature: sig(),
    }
}

#[test]
fn session_token_serialization_bounds_timestamps() {
    assert!(serde_json::to_string(&token(MAX_SAFE, MAX_SAFE)).is_ok());
    assert!(serde_json::to_string(&token(-MAX_SAFE, 0)).is_ok());
    assert!(serde_json::to_string(&token(MAX_SAFE + 1, 0)).is_err());
    assert!(serde_json::to_string(&token(0, -(MAX_SAFE + 1))).is_err());
}

#[test]
fn delegation_cert_serialization_bounds_timestamps() {
    assert!(serde_json::to_string(&cert(-MAX_SAFE, MAX_SAFE, vec![])).is_ok());
    assert!(serde_json::to_string(&cert(0, MAX_SAFE + 1, vec![])).is_err());
    assert!(serde_json::to_string(&cert(-(MAX_SAFE + 1), 0, vec![])).is_err());
}

#[test]
fn max_rate_constraint_serialization_bounds_count_and_window() {
    let ok = Constraint {
        kind: "max_rate".into(),
        count: 5,
        window_s: MAX_SAFE,
        ..Default::default()
    };
    assert!(serde_json::to_string(&cert(0, 1, vec![ok])).is_ok());

    let bad_window = Constraint {
        kind: "max_rate".into(),
        count: 5,
        window_s: MAX_SAFE + 1,
        ..Default::default()
    };
    assert!(serde_json::to_string(&cert(0, 1, vec![bad_window])).is_err());

    let bad_count = Constraint {
        kind: "max_rate".into(),
        count: MAX_SAFE + 1,
        window_s: 300,
        ..Default::default()
    };
    assert!(serde_json::to_string(&cert(0, 1, vec![bad_count])).is_err());
}

#[test]
fn proof_bundle_serialization_bounds_challenge_at_and_stream_seq() {
    let mut bundle = ProofBundle {
        agent_id: "a".into(),
        agent_pub_key: pubkey(),
        delegations: vec![cert(0, 1, vec![])],
        challenge: vec![0; 32],
        challenge_at: MAX_SAFE,
        challenge_sig: sig(),
        session_context: vec![],
        stream_id: vec![0; 32],
        stream_seq: MAX_SAFE,
    };
    assert!(serde_json::to_string(&bundle).is_ok());
    bundle.challenge_at = MAX_SAFE + 1;
    assert!(serde_json::to_string(&bundle).is_err());
    bundle.challenge_at = 0;
    bundle.stream_seq = MAX_SAFE + 1;
    assert!(serde_json::to_string(&bundle).is_err());
}
